src/membrane_proof.rs

```diff
@@ -93,14 +87,74 @@ fn mem_proof_server_url() -> String {
     }
 }
 
-fn get_hpos_config() -> Fallible<Config> {
+pub fn get_hpos_config() -> Result<Config> {
     let config_path = env::var("HPOS_CONFIG_PATH")?;
     let config_json = fs::read(config_path)?;
     let config: Config = serde_json::from_slice(&config_json)?;
     Ok(config)
 }
 
-async fn try_registration_auth(config: &Config, holochain_public_key: PublicKey) -> Fallible<()> {
+/// get the mem-proof needed to be used in this setup
+/// for the holo servers we will want to pass a read-only mem-proof
+/// for the holoports we should always expect a mem-proof, else return an error that will stop the core-app from getting installed
+pub async fn get_mem_proof() -> Result<HashMap<String, Arc<SerializedBytes>>> {
+    let mut mem_proof = HashMap::new();
+    if force_use_read_only_mem_proof() {
+        // This setting is mostly going to be used by the holo servers like mem-proof-server and match-server
+        let read_only_mem_proof = Arc::new(SerializedBytes::from(UnsafeBytes::from(vec![0])));
+        mem_proof.insert("core-app".to_string(), read_only_mem_proof.clone());
+        mem_proof.insert("holofuel".to_string(), read_only_mem_proof);
+    } else if let Ok(proof) = load_mem_proof_file() {
+        mem_proof.insert("core-app".to_string(), proof.clone());
+        mem_proof.insert("holofuel".to_string(), proof);
+    } else {
+        // Try again to get a mem-proof
+        match crate::membrane_proof::try_mem_proof_server_inner(None).await {
+            Ok(_) => {
+                let proof = load_mem_proof_file()?;
+                mem_proof.insert("core-app".to_string(), proof.clone());
+                mem_proof.insert("holofuel".to_string(), proof);
+            }
+            Err(e) => {
+                return Err(anyhow!(format!(
+                    "Unable to fetch a required mem-proof. {:?}",
+                    e
+                )))
+            }
+        }
+    }
+
+    Ok(mem_proof)
+}
+
+/// reads the mem-proof that is stored on the holoport
+/// this proof is used for the core-app i.e. hha and holofuel
+#[instrument(err)]
+pub fn load_mem_proof_file() -> Result<MembraneProof> {
+    use std::str;
+    let path = mem_proof_path();
+    let file = fs::read(&path).context("failed to open file")?;
+    let mem_proof_str = str::from_utf8(&file)?;
+    debug!("Loaded Proof {:?}", mem_proof_str);
+    let mem_proof_bytes = base64::decode(mem_proof_str)?;
+    let mem_proof_serialized = Arc::new(SerializedBytes::from(UnsafeBytes::from(mem_proof_bytes)));
+    Ok(mem_proof_serialized)
+}
+
+#[instrument(err, skip(holochain_public_key))]
+pub async fn try_mem_proof_server_inner(holochain_public_key: Option<PublicKey>) -> Result<()> {
+    let config = crate::membrane_proof::get_hpos_config()?;
+    let agent_pub_key = match holochain_public_key {
+        Some(a) => a,
```

Comment:
```suggestion
        Some(k) => k,
```
Small semantic suggestion.